services/wallet/router/filter.go

```diff
@@ -0,0 +1,154 @@
+func filterNetworkComplianceV2(routes [][]*PathV2, fromLockedAmount map[uint64]*hexutil.Big) [][]*PathV2 {
+	filteredRoutes := make([][]*PathV2, 0)
+	if routes == nil || fromLockedAmount == nil {
+		return filteredRoutes
+	}
+
+	fromIncluded, fromExcluded := setupRouteValidationMapsV2(fromLockedAmount)
+
+	for _, route := range routes {
+		if route == nil {
+			continue
+		}
+
+		// Create fresh copies of the maps for each route check, because they are manipulated
+		if isValidForNetworkComplianceV2(route, copyMap(fromIncluded), copyMap(fromExcluded)) {
+			filteredRoutes = append(filteredRoutes, route)
+		}
+	}
+	return filteredRoutes
+}
+
+// isValidForNetworkComplianceV2 checks if a route complies with network inclusion/exclusion criteria.
+func isValidForNetworkComplianceV2(route []*PathV2, fromIncluded, fromExcluded map[uint64]bool) bool {
+	logger.Debug("Initial inclusion/exclusion maps",
+		zap.Any("fromIncluded", fromIncluded),
+		zap.Any("fromExcluded", fromExcluded),
+	)
+
+	for _, path := range route {
+		if path == nil || path.From == nil {
+			logger.Debug("Invalid path", zap.Any("path", path))
+			return false
+		}
+		if _, ok := fromExcluded[path.From.ChainID]; ok {
+			logger.Debug("Excluded chain ID", zap.Uint64("chainID", path.From.ChainID))
+			return false
+		}
+		if _, ok := fromIncluded[path.From.ChainID]; ok {
+			fromIncluded[path.From.ChainID] = true
+		}
+	}
+
+	logger.Debug("fromIncluded after loop", zap.Any("fromIncluded", fromIncluded))
+
+	for chainID, included := range fromIncluded {
+		if !included {
+			logger.Debug("Missing included chain ID", zap.Uint64("chainID", chainID))
+			return false
+		}
+	}
+
+	return true
+}
+
+// setupRouteValidationMapsV2 initializes maps for network inclusion and exclusion based on locked amounts.
+func setupRouteValidationMapsV2(fromLockedAmount map[uint64]*hexutil.Big) (map[uint64]bool, map[uint64]bool) {
+	fromIncluded := make(map[uint64]bool)
+	fromExcluded := make(map[uint64]bool)
+
+	for chainID, amount := range fromLockedAmount {
+		if amount.ToInt().Cmp(zero) <= 0 {
+			fromExcluded[chainID] = false
+		} else {
+			fromIncluded[chainID] = false
+		}
+	}
+	return fromIncluded, fromExcluded
+}
+
+// filterCapacityValidationV2 performs the second level of filtering based on amount and capacity validation.
+func filterCapacityValidationV2(routes [][]*PathV2, amountIn *big.Int, fromLockedAmount map[uint64]*hexutil.Big) [][]*PathV2 {
+	filteredRoutes := make([][]*PathV2, 0)
+
+	for _, route := range routes {
+		if hasSufficientCapacityV2(route, amountIn, fromLockedAmount) {
+			filteredRoutes = append(filteredRoutes, route)
+		}
+	}
+	return filteredRoutes
+}
+
+// hasSufficientCapacityV2 checks if a route has sufficient capacity to handle the required amount.
+func hasSufficientCapacityV2(route []*PathV2, amountIn *big.Int, fromLockedAmount map[uint64]*hexutil.Big) bool {
+	for _, path := range route {
+		if amount, ok := fromLockedAmount[path.From.ChainID]; ok {
```

Comment:
@Samyoul the logic in this function should check if sum of amounts of all calculated paths meet the `amountIn` and I see that you're skipping paths on chain X if their amounts are not locked. This is misleading.

Also not sure why we have this complicate logic, I mean, calculating restAmountIn for other than this path and then comparing only if amount was added and...

A simple one could be (pseudo code):
```
function checkIfRouteMatchCriteriaToBeCandidateRoute(route, amountIn, fromLockedAmount) (bool, error) {
  routeTotal = 0
  for path in route {
    if (path.ChainId in fromLockedAmount && path.Amount != fromLockedAmount[path.ChainId] {
      return false, errors.New("amount on chain doesn't match the locked amount for this chain")
    }
    routeTotal += path.AmountIn
  }

  if (amountIn != routeTotal) {
    return false, errors.New("sum of amounts on all paths for this route doesn't meet amount in")
  }
  return true, nil
}
```

Reply:
AmountIn is set by the user, the routing algorithm doesn't change it. 
AmountIn per path and per chain is set either by the router algorithm if no locked value is provided for that chain or by the routing logic.
I don't see that anything is updated.

Reply:
The logic here is basically identical to the original code: 

https://github.com/status-im/status-go/blob/d2343ee03ec0277c92fb31001c36f8760d57d74f/services/wallet/router/router.go#L183-L208

I'm happy to consider improving the code in another PR but to ensure we have a clean refactor I'd be more comfortable with merging the original logic and then following up with additional changes.

Reply:
Ok @saledjenic and @IvanBelyakoff I've created a new issue for this specific issue. https://github.com/status-im/status-go/issues/5244. 

Please let me know if any details of this issue are incorrect or need additional information.

Reply:
This one is completely fine.